File: June 2022/June_19.java
class Solution {
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        Node root = new Node();
        
        for(String product : products){
            Node curr = root;
            for(char ch : product.toCharArray()){
                if(curr.children[ch - 'a'] == null)
                    curr.children[ch - 'a'] = new Node();
                curr = curr.children[ch - 'a'];
            }
            curr.isEnd = true;
        }
        
        List<List<String>> list = new ArrayList<>();
        Node curr = root;
        
        StringBuilder product  = new StringBuilder();
        for(char ch : searchWord.toCharArray()){
            if(curr.children[ch - 'a'] == null)
                break;
            curr = curr.children[ch - 'a'];
            
            product.append(ch);
            List<String> li = new ArrayList<>();
            getProducts(curr, li, product);
            list.add(li);
        }
        
        while(list.size() < searchWord.length())
            list.add(new ArrayList<>());
        return list;
    }
    
    public void getProducts(Node curr, List<String> li, StringBuilder product){
        if(li.size() == 3)
            return;
        
        if(curr.isEnd)
            li.add(product.toString());
        
        for(int i = 0; i < 26; i++)
            if(curr.children[i] != null){
                product.append((char)('a' + i));
                getProducts(curr.children[i], li, product);
                product.deleteCharAt(product.length() - 1);
            }
    }
    
    class Node{
        boolean isEnd;
        Node[] children;
        Node(){
            isEnd = false;
            children = new Node[26];
        }
    }
}

File: June 2022/June_19.cpp
class Solution {
public:
    class Node{
        public:
        bool isEnd;
        vector<Node*> children{vector<Node*>(26, nullptr)};
        Node(){
            isEnd = false;
        }
    };
    
    void getProducts(Node* curr, vector<string> &v, string &product){
        if(v.size() == 3)
            return;
        
        if(curr->isEnd)
            v.push_back(product);
        
        for(int i = 0; i < 26; i++)
            if(curr->children[i] != nullptr){
                product += (char)('a' + i);
                getProducts(curr->children[i], v, product);
                product.pop_back();
            }
    }
    
    vector<vector<string>> suggestedProducts(vector<string>& products, string searchWord) {
        Node* root = new Node();
        for(string &product : products){
            Node* curr = root;
            for(char &ch : product){
                if(curr->children[ch - 'a'] == nullptr)
                    curr->children[ch - 'a'] = new Node();
                curr = curr->children[ch - 'a'];
            }
            curr->isEnd = true;
        }
        
        vector<vector<string>> vec(searchWord.length(), vector<string>());
        string product = "";
        Node* curr = root;
        for(int i = 0; i < searchWord.length(); i++){
            char ch = searchWord[i];
            if(curr->children[ch - 'a'] == nullptr)
                break;
            curr = curr->children[ch - 'a'];
            product += ch;
            getProducts(curr, vec[i], product);
        }
        return vec;
    }
};
